File: gcpc2019/insertionorder/submissions/accepted/tobias.py
#!/usr/bin/env python3
# -*- coding: UTF-8 -*-

import sys


def create_tree_balanced(start, end):
    if end < start:
        return
    mid = (start + end) // 2
    print(mid, end=" ")
    create_tree_balanced(start, mid - 1)
    create_tree_balanced(mid + 1, end)


def create_tree(start, end, depth_rem):
    if depth_rem <= 0:
        create_tree_balanced(start, end)
    else:
        mid = (start + end) // 2
        first = max(start, mid - depth_rem)
        for i in range(first, mid + 1):
            print(i, end=" ")
        create_tree_balanced(start, first - 1)
        create_tree(mid + 1, end, depth_rem - mid + first)


if __name__ == "__main__":
    n, k = map(int, sys.stdin.readline().split())

    i = n
    while i:
        k -= 1
        i //= 2
    if k < 0:
        print("impossible")
    else:
        create_tree(1, n, k)
        print()

File: gcpc2019/insertionorder/submissions/accepted/Tobias.java
import java.io.*;
import java.nio.charset.StandardCharsets;

public class Tobias {
    private static final StringBuilder builder = new StringBuilder(8096);
    private static void createTreeBalanced(int start, int end) {
        if (end < start) {
            return;
        }
        int mid = (start + end) >>> 1;
        builder.append(mid).append(" ");
        createTreeBalanced(start, mid - 1);
        createTreeBalanced(mid + 1, end);
    }

    private static void createTree(int start, int end, int depth) {
        if (depth <= 0) {
            createTreeBalanced(start, end);
        } else {
            int mid = (start + end) >>> 1;
            int first = Math.max(start, mid - depth);
            for (int i = first; i <= mid; i++) {
                builder.append(i).append(" ");
            }
            createTreeBalanced(start, first - 1);
            createTree(mid + 1, end, depth - mid + first);
        }
    }

    public static void main(String... args) throws IOException {
        try (BufferedReader reader = new BufferedReader(
                new InputStreamReader(System.in, StandardCharsets.UTF_8))) {
            String[] data = reader.readLine().split(" ");
            int n = Integer.parseInt(data[0]);
            int k = Integer.parseInt(data[1]);

            for (int i = n; i > 0; i /= 2) {
                k--;
            }
            if (k < 0) {
                System.out.println("impossible");
            } else {
                createTree(1, n, k);
                System.out.println(builder.toString());
            }
        }
    }
}
